test/unittests/validation_test.cpp

```diff
@@ -231,3 +231,68 @@ TEST(validation, call_indirect_no_table)
     const auto wasm = from_hex("0061736d0100000001050160017f00030201000a0901070020001101000b");
     EXPECT_THROW_MESSAGE(parse(wasm), validation_error, "call_indirect without defined table");
 }
+
+TEST(validation, export_invald_index)
+{
+    /* wat2wasm --no-check
+    (export "foo" (func 0))
+    */
+    const auto wasm_func = from_hex("0061736d0100000007070103666f6f0000");
+    EXPECT_THROW_MESSAGE(
+        parse(wasm_func), validation_error, "invalid index of an exported function");
+
+    /* wat2wasm --no-check
+    (export "g" (global 0))
+    */
+    const auto wasm_glob = from_hex("0061736d0100000007050101670300");
+    EXPECT_THROW_MESSAGE(parse(wasm_glob), validation_error, "invalid index of an exported global");
+
+    /* wat2wasm --no-check
+    (export "t" (table 0))
+    */
+    const auto wasm_no_table = from_hex("0061736d0100000007050101740100");
+    EXPECT_THROW_MESSAGE(
+        parse(wasm_no_table), validation_error, "invalid index of an exported table");
+
+    /* wat2wasm --no-check
+    (table 1 funcref)
+    (export "t" (table 1))
+    */
+    const auto wasm_table = from_hex("0061736d0100000004040170000107050101740101");
+    EXPECT_THROW_MESSAGE(parse(wasm_table), validation_error, "invalid index of an exported table");
+
+    /* wat2wasm --no-check
+    (export "m" (memory 0))
+    */
+    const auto wasm_no_mem = from_hex("0061736d01000000070501016d0200");
+    EXPECT_THROW_MESSAGE(
+        parse(wasm_no_mem), validation_error, "invalid index of an exported memory");
+
+    /* wat2wasm --no-check
+    (memory 1)
+    (export "m" (memory 1))
+    */
+    const auto wasm_mem = from_hex("0061736d010000000503010001070501016d0201");
+    EXPECT_THROW_MESSAGE(parse(wasm_mem), validation_error, "invalid index of an exported memory");
```

Comment:
Should we have different message whether memory/table section is present or not?

Reply:
I think this is fine - 0 is invalid index in some context.